src/stores/undoStore.ts

```diff
@@ -0,0 +1,42 @@
+
+export class UndoStore<UndoObject> {
+    private _undoObjects: UndoObject[];
+    private _undoIndex: number;
+
+    constructor() {
+        this.clear();
+    }
+
+    public addUndoObject = (undoObject: UndoObject) => {
+        // Remove the history of undo's because current state is changed
+        this._undoObjects.splice(this._undoIndex + 1);
+
+        // Insert current undoObject to the pile
```

Comment:
Would suggest the word stack, a stack is ordered, a pile is not necessarily ordered.